### ISSUE TITLE
unnecessary-optional-chain lint triggers incorrectly for array indexer access

### ISSUE THREAD

**Issue by fizker** [facebook/flow#8187]
When accessing values of an array of type `Array<T>`, `undefined` may be returned. But flow insists that they are always type `T`. I understand why this works this way and I am mostly ok with that (from a pure soundness perspective, it is not ok. But that is a different discussion).

The issue is that the `unnecessary-optional-chain` lint option triggers for defensive code that is actually necessary. 

In the example below, a function accesses a property on the first item in the array. If the array is empty and we followed the flow recommendation, this would fail with a null-exception.

```
// flowlint unnecessary-optional-chain:error

function getFirst(x:$ReadOnlyArray<{| foo: string |}>) : ?string {
  return x[0]?.foo
}
```

I know there are multiple workarounds, I would just prefer that I did not have to use them.

Please note that this is not about making flow insist that `array[index]` returns an optional value, it is only about the linter accepting that this is a perfectly valid place to use optional chaining.

Flow version: 0.112.0

## Expected behavior

The linter should not trigger.

## Actual behavior

The `unnecessary-optional-chain` lint triggers.

https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVxhlomBLAOwBcwBXAggUwGNKBnOgQwCcBPAWjgAci84DGMdtQAWjQgC5KzZnGboo5ar35gA5pSIAxPMzpEAFAA8JAEgBKlRgBMA8gRisAgjMasAPAG8APtjhwJMH1mQjUwbwBfAD4ASjBAgH5g0LBPVDAwZk1SZgIwIwBtAAYAXQSAOih-VAjUIA

**Comment by Brianzchen** [comment]
Was discussed in flow discord channel. Posting here because related. Have edited with some formatting to make it readable.

---

**TrySound 10/03/2020**
Hey guys, I have enabled unnecessary-optional-chaining lint rule and struggle with accessing array by index. Array items can be undefined. And here lint rule tells array[0] is not nullable. Can array index be an exception for this lint rule?
```
const array: Array<{| a: 1 |}> = []
array[0]?.a
```
**malectro 10/03/2020**
beyond the lint rule, it's always worried me that Flow considers indexing with a literal to yield a defined value

**Nat 10/03/2020**
setting aside the lint rule for a moment, the array indexer return type is a long-standing debate...
let me see if I can dig up a thread
this will likely provide some useful context: https://github.com/facebook/flow/pull/6823
the summary is that Flow decided long ago to accept some unsoundness here for the sake of ease-of-use
the behavior of array indexers in JS leaves us with no good options here
if I were designing JS arrays, I would probably have included a bounds check and made them throw if the array index were out of bounds. that would allow Flow to safely assume that if an array access returns, it returns a non-nullable value (if, of course, the array contents are non-nullable)
unfortunately, because of the way JS arrays behave, we have to choose between two bad options:

- Force people to check all of their array accesses for null, or
- Unsoundly assume array accesses are in-bounds

the first option is especially bad because people often write their own bounds checks, but Flow does not have the machinery to analyze bounds-checks to conclude that the indices are in-bounds, e.g. in a for loop iterating through the indices of an array
so, in essence people would be forced to check twice
we are moving towards soundness in general, but we have to pick our battles. we believe this change would be pretty unpopular, and we know it would have a significant rollout burden. we may make it eventually, but we don't have any short-term plans to do so
now, to address the lint rule question:
because of the above, as far as Flow is concerned, this is not nullable. we don't have any mechanism by which to indicate that this particular type is untrusted (though we have explored doing a trust analysis, and will likely have something along those lines eventually). so, for now, there's no way to selectively disable the lint for cases like this, and I don't see a reasonable short-term path to such functionality
I know this isn't the answer you were hoping for...
sainati might also have some thoughts on this

**sainati 10/03/2020**
my suggestion is to add a `// $FlowIgnore[unnecessary-optional-chain]` comment above that line
but yes Nat summarized the issue well and atm don't expect anything to change here

**TrySound 10/03/2020**
We currently do length check instead of optional chaining to not hide other errors.
But it's forgotten sometimes

**sainati 10/03/2020**
you won't hide other errors if you use this suppression, that will only suppress this lint specifically

**TrySound 10/03/2020**
I mean other chaining errors

**malectro 10/03/2020**
obviously you guys have thought about this way more than i have so this is probably silly to ask, but was there any discussion around the difference between `array[0]` and `array[i]`? the first is something i'd always assume to be possibly undefined – to the point where i wrote a custom `first<V>(Array<V>): V | void`  utility in our code base to use instead.

**sainati 10/03/2020**
ah i see

**TrySound 10/03/2020**
Btw, I saw proposal about `array.item(i)`. Can it be nullable like `map.get(key)`?

**Nat 10/03/2020**
malectro the fundamental issues at play aren't really different. your utility is a reasonable thing to add
TrySound I'm not aware of that, but if it's something new I think we could make it nullable
https://github.com/tc39/proposal-item-method
nice, stage 3
yeah, when we add that I think we should add it with a nullable return value
